Update command

alenkacz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/kudobuilder/kudo/blob/master/CONTRIBUTING.md
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://github.com/kudobuilder/kudo/blob/master/RELEASE.md
3. Ensure you have added or ran the appropriate tests for your PR
4. If the PR is unfinished, start it as a Draft PR: https://github.blog/2019-02-14-introducing-draft-pull-requests/
-->

**What type of PR is this?**
/component kudoctl
/kind enhancement

**What this PR does / why we need it**:
`kudo update` patches running instance with new parameter values.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Added `kudo update` command to change parameters of running operator.
```

jbarrick-mesosphere commented on pkg/kudoctl/cmd/update.go:
```diff
@@ -0,0 +1,108 @@
+			}
+			return runUpdate(args, options)
+		},
+		SilenceUsage: true,
```
Why are we setting this?

alenkacz replied:
lol good question :D I don't know. Copy-paste 😛 

`SilenceUsage is an option to silence usage when an error occurs.` ... sounds reasonable to me though 🤷‍♀ lot of errors can come up from non-usage errors so maybe that's why we have decided to not print it there? No hard opinion though...

alenkacz replied:
okay, should I remove it then from all comands?

zen-dog commented on pkg/kudoctl/cmd/update.go:
```diff
@@ -0,0 +1,108 @@
+	return nil
+}
+
+func runUpdate(args []string, options *updateOptions) error {
```
We have this patter (which I like) where we separate building of the command from the actual code that is executed in different files. It might seem like overkill here, but these things tend to grow 😉

alenkacz replied:
yeah my current guidance is to keep it in one file as long as the command is simple enough...

zen-dog replied:
...for now 😉 But I'll leave this up to you

zen-dog commented on pkg/kudoctl/cmd/upgrade.go:
```diff
@@ -157,7 +158,7 @@ func upgrade(newOv *v1alpha1.OperatorVersion, kc *kudo.Client, options *options)
 
 	// Change instance to point to the new OV and optionally update parameters
-	err = kc.UpdateInstance(options.InstanceName, options.Namespace, newOv.Name, options.Parameters)
+	err = kc.UpdateInstance(options.InstanceName, options.Namespace, util.String(ov.Name), options.Parameters)
```
Hu? Why did this change from `newOv` to `ov`?

alenkacz replied:
good catch :)

zen-dog commented on pkg/kudoctl/cmd/update.go:
```diff
@@ -0,0 +1,109 @@
+		return fmt.Errorf("instance %s in namespace %s does not exist in the cluster", instanceToUpdate, options.Namespace)
+	}
+
+	// Change instance to point to the new OV and optionally update parameters
```
I don't think this comment makes sense here

zen-dog commented on pkg/kudoctl/cmd/update.go:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	// Change instance to point to the new OV and optionally update parameters
+	err = kc.UpdateInstance(instanceToUpdate, options.Namespace, nil, options.Parameters)
```
also, why do we pass `nil` as `operatorVersionName` here?

alenkacz replied:
because we don't want to update the version in this call. I use nil here as I would use `None` in scala

zen-dog commented on pkg/kudoctl/cmd/update_test.go:
```diff
@@ -0,0 +1,99 @@
+		errMessageContains string
+	}{
+		{"instance does not exist", false, map[string]string{"param": "value"}, "instance test in namespace default does not exist in the cluster"},
+		{"update parameters", true, map[string]string{"param": "value"}, ""},
```
Nice tests 👏 How about giving the `testInstance` an existing parameter `param: oldValue`?

alenkacz replied:
yeah I think the current patch implementation is incorrect, since this code for patch is already on master, I'll address that in another PR

alenkacz replied:
okay I did it in here

alenkacz commented:
okay, the patch for parameters was not working 100% because if you had parameters in place, it dropped them and replaced with new, which is not correct. I switched to merge patch and now it seems to be OK

zen-dog commented on pkg/kudoctl/cmd/update.go:
```diff
@@ -0,0 +1,109 @@
+	updateCmd := &cobra.Command{
+		Use:     "update <instance-name>",
+		Short:   "Update installed KUDO operator.",
+		Long:    `Update installed KUDO operator with new parameters.`,
```
Should we also mention explicitly that one can not remove existing parameters this way?
